reasoner/planner/GreedyCostSearch.java

```diff
@@ -0,0 +1,102 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+package com.vaticle.typedb.core.reasoner.planner;
+
+import com.vaticle.typedb.core.concept.ConceptManager;
+import com.vaticle.typedb.core.logic.LogicManager;
+import com.vaticle.typedb.core.logic.resolvable.Resolvable;
+import com.vaticle.typedb.core.logic.resolvable.ResolvableConjunction;
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public abstract class GreedyCostSearch extends ReasonerPlanner {
+
+    public GreedyCostSearch(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
+        super(traversalEng, conceptMgr, logicMgr);
+    }
+
+    /* TODO:
+     * AnswerSize -> For each resolvable: Add the number of calls + the answers produced
+     * AnswersProduced? TraversalEngine tells us this based on constraints in the rule-body
+     * #calls? Product of the cardinality of each variable in the bounds (This number reduces as you evaluate a new resolvable which further constrains the variable - but min or less?)ssssssss
+     * Effect? Schedule stronger constraints first.
+     */
+    @Override
+    protected Plan<Resolvable<?>> planConjunction(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> inputBounds) {
+        Set<Identifier.Variable.Retrievable> bounds = new HashSet<>(inputBounds);
+        Set<Resolvable<?>> resolvables = compile(conjunction);
+
+        Set<Resolvable<?>> remaining = new HashSet<>(resolvables);
+        long cost = 0;
+        List<Resolvable<?>> orderedResolvables = new ArrayList<>();
+        Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies = dependencies(resolvables);
+
+        while (!remaining.isEmpty()) {
+            Optional<Resolvable<?>> nextResolvableOpt = remaining.stream()
+                    .filter(r -> dependenciesSatisfied(r, bounds, dependencies))
+                    .min(Comparator.comparing(r -> estimateCost(r, bounds)));
+
+            if (nextResolvableOpt.isEmpty()) {
+                nextResolvableOpt = remaining.stream()
+                        .filter(r -> !dependenciesSatisfied(r, bounds, dependencies) && !r.isNegated())
+                        .min(Comparator.comparing(r -> estimateCost(r, bounds)));
+            }
+            assert nextResolvableOpt.isPresent();
+            Resolvable<?> nextResolvable = nextResolvableOpt.get();
+            cost += estimateCost(nextResolvable, bounds); // TODO: Eliminate double work
```

Comment:
I think we should implement this TODO, we can discuss how we may be able to do that if it's not easy!
Note it's not just this place we do duplicate work: in the first `stream()` we may be doing the same resolvable + bounds multiple times.

Reply:
not sure that itself is solvable without a good equality function on Retrievables so maybe we just don't worry about it for now

Reply:
That's ok. We can use object equality.
But this will probably be redundant as soon as we update the search.